src/cmd/services/m3coordinator/downsample/id_pool_types.go

```diff
@@ -0,0 +1,327 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package downsample
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/m3db/m3db/src/dbnode/serialize"
+	"github.com/m3db/m3metrics/metric/id"
+	"github.com/m3db/m3x/checked"
+	"github.com/m3db/m3x/ident"
+	"github.com/m3db/m3x/pool"
+
+	"github.com/prometheus/common/model"
+)
+
+var (
+	metricNameTagName = []byte(model.MetricNameLabel)
+	rollupTagName     = []byte("m3_rollup")
+	rollupTagValue    = []byte("true")
+
+	errNoMetricNameTag = errors.New("no metric name tag found")
+)
+
+// Ensure encodedTagsIterator implements id.SortedTagIterator
+var _ id.SortedTagIterator = &encodedTagsIterator{}
+
+// Ensure encodedTagsIterator implements id.ID
+var _ id.ID = &encodedTagsIterator{}
+
+type encodedTagsIterator struct {
+	tagDecoder serialize.TagDecoder
+	bytes      checked.Bytes
+	pool       *encodedTagsIteratorPool
+}
+
+func newEncodedTagsIterator(
+	tagDecoder serialize.TagDecoder,
+	pool *encodedTagsIteratorPool,
+) *encodedTagsIterator {
+	return &encodedTagsIterator{
+		tagDecoder: tagDecoder,
+		bytes:      checked.NewBytes(nil, nil),
```

Comment:
Since `encodedTagsIterator` itself is already pooled this is only paid when a single `encodedTagsIterator` is created in the pool so I think it would be the same cost to also pool the `checked.NewBytes(...)` because the entire `encodedTagsIterator` will be reused anyhow, hence we don't allocate `checked.NewBytes(...)` over and over.